### ISSUE TITLE
Usability issues with error/notice/success messages

### ISSUE THREAD

**Issue by erikhansen** [magento/magento2#2916]
This issue covers what I consider to be two major usability issues with messages:
#1. Error message hides after 5 seconds

When error messages are displayed on the frontend, they disappear after 5 seconds (this is controlled by the `onHiddenChange` method in `app/code/Magento/Ui/view/frontend/web/js/view/messages.js`). If a user is not paying attention, if they are a slow reader, or if the error message is long, they message may disappear before they read the entire message. My recommendation would be to NOT hide the message but to leave it on the page until the user takes some other action that causes the message to be hidden.

Here is an example of an error message that will hide after 5 seconds:

![image](https://cloud.githubusercontent.com/assets/129031/12174181/bb6a3f60-b520-11e5-8b31-f10eac623589.png)

@mttjohnson did some digging into what happens when there is an error with a payment method in the checkout process. It looks like the payment methods use the `Magento_Ui/js/modal/alert` component to display an alert-style modal that doesn't go away until the user closes it. The red error message I displayed above occurs when there is an exception thrown by Magento when progressing from Shipping > Billing, so it may not warrant an alert style popup, but it should at least persist on the page and not hide. This is what the alert-style popup looks like:

![image](https://cloud.githubusercontent.com/assets/129031/12175502/486820e2-b528-11e5-9b08-66afdcf8c617.png)
#2. Error messages may display outside of current viewport

If a user is a on a small screen device, there are many situations where they will never see an error message. For example, if they're on the "Shipping" step of the checkout process and an error message is added, they may not see the error message get added to the top of the page. My suggestion would be to either smooth scroll the user up to the error message, display the error message next to the button they used to submit, or come up with a notification system that display notifications on top of all content.

![checkout](https://cloud.githubusercontent.com/assets/129031/12174273/411d1038-b521-11e5-91de-30e10434a262.png)

**Comment by MomotenkoNatalia** [comment]
Hi @erikhansen  
We have created task for our Design team to refine a pattern of messages from your proposals - Internal ticket MAGETWO-47763

Thanks for sharing your ideas!

**Comment by tkacheva** [comment]
Implementation task MAGETWO-54422

**Comment by ErvinLlojku** [comment]
Hello, 

Please kindfully consider updating the error messages for the swatch elements. It is the same usability issue as described above. Should be reconsidered the way the styles are applyied to the swatch inputs.

Thank you!

**Comment by tkacheva** [comment]
@ErvinLlojku could you provide more details about swatch error messages please. I suggest to open separate ticket for this

**Comment by blizam** [comment]
This is huge I think... especially on credit card processing errors... if a card is declined in any of the PayPal checkout options there's a good chance the error is outside of the users' window (above) and because the error clears so quickly (it hides immediately if you put your mouse on it; or after 5 seconds) the user will quite often miss it (I know I did in testing) and therefore have no idea if their payment went through or what they should do next.

The error messages should persist on checkout and not auto-hide

**Comment by erikhansen** [comment]
@blizam Yes, I consider this a serious usability issue and I hope Magento addresses it in the next release.

**Comment by adragus-inviqa** [comment]
Hijacking this thread to express my facepalming: I don't know whose idea was to hide the messages after 5 seconds, but it's a darn bad decision.

**Comment by piotrekkaminski** [comment]
Thank you for your submission.

We recently made some changes to the way we process GitHub submissions to more quickly identify and respond to core code issues.

Feature Requests and Improvements should now be submitted to the new [Magento 2 Feature Requests and Improvements](https://community.magento.com/t5/Magento-2-Feature-Requests-and/idb-p/feature-requests) forum (see details [here](https://community.magento.com/t5/News-Announcements/Improvements-to-GitHub-Management/m-p/44572#M96)).

We are closing this GitHub ticket and have moved your request to the new [forum](https://community.magento.com/t5/Magento-2-Feature-Requests-and/Usability-issues-with-error-notice-success-messages/idi-p/46431).

**Comment by erikhansen** [comment]
@MomotenkoNatalia I posted a comment to you in the forums: https://community.magento.com/t5/Magento-2-Feature-Requests-and/Usability-issues-with-error-notice-success-messages/idc-p/53406#M2548

**Comment by Sanfam** [comment]
@MomotenkoNatalia , I, much like @erikhansen , am eager to know where Magento stands on this issue.  I've posted a comment in the forum but noticed that Magento has been silent on the topic since August 2016.

**Comment by tuyennn** [comment]
I found no go with this issue, no way to fix this after trying so much Temporary fixes

**Comment by cykirsch** [comment]
@magento-engcom-team I don't see any referenced commits. So no specific changes were made, you just noticed it was fixed in 2.3? Can you figure out what the difference is and backport to 2.2 and 2.1? This is really important and several people in both this thread and in the forum thread have made excellent points describing its importance.

**Comment by cykirsch** [comment]
@MomotenkoNatalia

**Comment by erikhansen** [comment]
I patched `magento/module-ui` using the [composer-patches](https://github.com/cweagans/composer-patches) plugin to show the error message for 30 seconds, and scroll it into view. 

Here are the contents of the patch:

```
From b56d22d93e24b45a2ecbcf9f849502411cb99829 Mon Sep 17 00:00:00 2001
From: Erik Hansen
Date: Wed, 16 Nov 2016 16:50:31 -0600
Subject: [PATCH] Show error message for 30 seconds and scroll it into view

---
 view/frontend/web/js/view/messages.js | 7 ++++++-
 1 file changed, 6 insertions(+), 1 deletion(-)

diff --git a/view/frontend/web/js/view/messages.js b/view/frontend/web/js/view/messages.js
index e6d3bc8..ddb9a99 100644
--- a/view/frontend/web/js/view/messages.js
+++ b/view/frontend/web/js/view/messages.js
@@ -50,7 +50,12 @@ define([
             if (isHidden) {
                 setTimeout(function () {
                     $(self.selector).hide('blind', {}, 500)
-                }, 5000);
+                }, 30000);
+                try {
+                    $(self.selector)[0].scrollIntoView();
+                } catch (error) {
+                    // Do nothing
+                }
             }
         }
     });
-- 
2.8.1

```

**Comment by djview** [comment]
If somebody doesn't want to run the composer patch above, you can make this change by editing the js file directly by going to vendor/magento/module-ui/view/frontend/web/js/view/messages.js and editing starting at line 67. Delete the following code:

if (isHidden) {
                setTimeout(function () {
                    $(self.selector).hide('blind', {}, 500);
                }, 5000);
            }

And replace it with:

            if (isHidden) {
                 setTimeout(function () {
                     $(self.selector).hide('blind', {}, 500);
                }, 30000);
                try {
                    $(self.selector)[0].scrollIntoView();
                } catch (error) {
                    // Do nothing
                }
             }

**Comment by vovayatsyuk** [comment]
@erikhansen, thanks for your patch.

By the way, I'm interested why the following code having a place in [view/messages.js](https://github.com/magento/magento2/blob/2.2-develop/app/code/Magento/Ui/view/frontend/web/js/view/messages.js#L50-L52):

```js
        isVisible: function () {
            return this.isHidden(this.messageContainer.hasMessages());
        },
```

The problem is that this method **always returns js object** instead of boolean value and changes `isHidden` value to true when error messages found.

I don't know, but for me, it looks like an accident error.

It was [added a long time ago](https://github.com/magento/magento2/commit/9fabd4599017e727acaad6c101b851a5881f6400#diff-88469bcdfc33c043ae311e3ee4b4624eL21) without any explanation.

p.s. If we change this method as follows, messages will not we hidden automatically anymore:

```js
        isVisible: function () {
            return this.messageContainer.hasMessages();
        },
```

**Comment by ywsw** [comment]
@djview @erikhansen  thank you so much for your fix. It definitely works for me - only that the page scrolls up to the error notice on the first error only. For any consecutive errors, the page does not scroll back to the error (though the error notice does still last longer - if the user decides to scroll back up to the top). Would you be able to advise how this can be fixed?

**Comment by dawhoo** [comment]
> @djview @erikhansen thank you so much for your fix. It definitely works for me - only that the page scrolls up to the error notice on the first error only. For any consecutive errors, the page does not scroll back to the error (though the error notice does still last longer - if the user decides to scroll back up to the top). Would you be able to advise how this can be fixed?

Try this:

`        if (isHidden) {
             setTimeout(function () {
                 $(self.selector).hide('blind', {}, 500);
            }, 30000);
            try {
                $(self.selector)[0].scrollIntoView(alignToTop);
            } catch (error) {
                // Do nothing
            }
         }`

It seems to have worked for me.

**Comment by avra911** [comment]
Maybe 
```
 $("html, body").animate({scrollTop: 0}, "slow");
```
works better than 
```
$(self.selector)[0].scrollIntoView();
```

for consecutive errors.

**Comment by przmak** [comment]
This is still not fixed, my messages are dispersing, 4 years a show success/error message functionality does not work....  how hard can that be?

**Comment by junnhouse** [comment]
This fixed only displaying on Safari.  Chrome and Firefox is not working

**Comment by christianbookpwood** [comment]
Moving this to the forum doesn't seem to have helped, since the issue still isn't fixed. Any idea if this is being worked on?

**Comment by erikhansen** [comment]
@christianbookpwood I'd recommend applying the patch I mentioned [in this comment](https://github.com/magento/magento2/issues/2916#issuecomment-345343749).

**Comment by djview** [comment]
I thought I'd let people know that I tried a different solution for this problem and it worked really well. I found some code to implement a modal popup box for the error message. You just need to drop a new file 

app/design/frontend/YOUR/TEMPLATE/Magento_Payment/templates/transparent/iframe.phtml

into your theme directory structure there. Copy paste the code from the top box at the following link and it worked for me out of the box:

https://magento.stackexchange.com/questions/270039/checkout-errors-inside-modal
